In [2]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder
import pandas as pd
import tensorflow as tf

# Import our input dataset
attrition_df = pd.read_csv('HR-Employee-Attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

**Вместо того, чтобы искать во всех 35 столбцах и отслеживать, какие переменные нуждаются в категориальной предварительной обработке, мы позволим Python выполнять всю тяжелую работу. В качестве дополнительного бонуса мы можем использовать наш список переменных для выполнения одноразовой кодировки один раз, а не для каждой отдельной переменной. Для создания нашего списка переменных мы будем использовать свойство Pandas Dataframe.dtypes.**

In [3]:
# Generate our categorical variable list
attrition_cat = attrition_df.dtypes[attrition_df.dtypes == "object"].index.tolist()

**Looking at our attrition_cat variable, there are eight categorical variables that need encoding. However, before we loop through our variables and encode them using Scikit-learn's OneHotEncoder module, we need to make sure that none of the categorical variables have more than 10 unique values and require bucketing. To check the three variables, add and run the following code:**

In [4]:
# Check the number of unique values in each column
attrition_df[attrition_cat].nunique()

Attrition         2
BusinessTravel    3
Department        3
EducationField    6
Gender            2
JobRole           9
MaritalStatus     3
Over18            1
OverTime          2
dtype: int64

**Согласно nunique методу, ни одна из категориальных переменных не имеет более 10 уникальных значений, что означает, что мы готовы кодировать. Ранее мы кодировали одну переменную за раз, но Scikit-learn достаточно гибок, чтобы выполнять все одногорячие кодировки одновременно. Единственное отличие от нашего примера одной переменной заключается в том, что нам нужно передать наш список переменных attrition_cat. Опять же, добавьте и запустите следующий код в свой блокнот:**

In [5]:
# Create a OneHotEncoder instance
enc = OneHotEncoder(sparse=False)

# Fit and transform the OneHotEncoder using the categorical variable list
encode_df = pd.DataFrame(enc.fit_transform(attrition_df[attrition_cat]))

# Add the encoded variable names to the DataFrame
encode_df.columns = enc.get_feature_names(attrition_cat)
encode_df.head()

/Users/dang/opt/anaconda3/envs/mlenv/lib/python3.7/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


Attrition_No  Attrition_Yes  BusinessTravel_Non-Travel  \
0           0.0            1.0                        0.0   
1           1.0            0.0                        0.0   
2           0.0            1.0                        0.0   
3           1.0            0.0                        0.0   
4           1.0            0.0                        0.0   

   BusinessTravel_Travel_Frequently  BusinessTravel_Travel_Rarely  \
0                               0.0                           1.0   
1                               1.0                           0.0   
2                               0.0                           1.0   
3                               1.0                           0.0   
4                               0.0                           1.0   

   Department_Human Resources  Department_Research & Development  \
0                         0.0                                0.0   
1                         0.0                                1.0   
2                         0.0                                1.0   
3                         0.0                                1.0   
4                         0.0                                1.0   

   Department_Sales  EducationField_Human Resources  \
0               1.0                             0.0   
1               0.0                             0.0   
2               0.0                             0.0   
3               0.0                             0.0   
4               0.0                             0.0   

   EducationField_Life Sciences  ...  JobRole_Research Director  \
0                           1.0  ...                        0.0   
1                           1.0  ...                        0.0   
2                           0.0  ...                        0.0   
3                           1.0  ...                        0.0   
4                           0.0  ...                        0.0   

   JobRole_Research Scientist  JobRole_Sales Executive  \
0                         0.0                      1.0   
1                         1.0                      0.0   
2                         0.0                      0.0   
3                         1.0                      0.0   
4                         0.0                      0.0   

   JobRole_Sales Representative  MaritalStatus_Divorced  \
0                           0.0                     0.0   
1                           0.0                     0.0   
2                           0.0                     0.0   
3                           0.0                     0.0   
4                           0.0                     0.0   

   MaritalStatus_Married  MaritalStatus_Single  Over18_Y  OverTime_No  \
0                    0.0                   1.0       1.0          0.0   
1                    1.0                   0.0       1.0          1.0   
2                    0.0                   1.0       1.0          0.0   
3                    1.0                   0.0       1.0          0.0   
4                    1.0                   0.0       1.0          1.0   

   OverTime_Yes  
0           1.0  
1           0.0  
2           1.0  
3           1.0  
4           0.0  

[5 rows x 31 columns]

**Теперь, когда наши категориальные переменные закодированы, они готовы заменить наши некодированные категориальные переменные в нашем наборе данных.**

**To replace these columns, we'll use a combination of Pandas' merge and drop methods. Add and run the following code in the notebook:**

In [6]:
# Merge one-hot encoded features and drop the originals
attrition_df = attrition_df.merge(encode_df,left_index=True, right_index=True)
attrition_df = attrition_df.drop(attrition_cat,1)
attrition_df.head()

/Users/dang/opt/anaconda3/envs/mlenv/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  This is separate from the ipykernel package so we can avoid doing imports until


Age  DailyRate  DistanceFromHome  Education  EmployeeCount  EmployeeNumber  \
0   41       1102                 1          2              1               1   
1   49        279                 8          1              1               2   
2   37       1373                 2          2              1               4   
3   33       1392                 3          4              1               5   
4   27        591                 2          1              1               7   

   EnvironmentSatisfaction  HourlyRate  JobInvolvement  JobLevel  ...  \
0                        2          94               3         2  ...   
1                        3          61               2         2  ...   
2                        4          92               2         1  ...   
3                        4          56               3         1  ...   
4                        1          40               3         1  ...   

   JobRole_Research Director  JobRole_Research Scientist  \
0                        0.0                         0.0   
1                        0.0                         1.0   
2                        0.0                         0.0   
3                        0.0                         1.0   
4                        0.0                         0.0   

   JobRole_Sales Executive  JobRole_Sales Representative  \
0                      1.0                           0.0   
1                      0.0                           0.0   
2                      0.0                           0.0   
3                      0.0                           0.0   
4                      0.0                           0.0   

   MaritalStatus_Divorced  MaritalStatus_Married  MaritalStatus_Single  \
0                     0.0                    0.0                   1.0   
1                     0.0                    1.0                   0.0   
2                     0.0                    0.0                   1.0   
3                     0.0                    1.0                   0.0   
4                     0.0                    1.0                   0.0   

   Over18_Y  OverTime_No  OverTime_Yes  
0       1.0          0.0           1.0  
1       1.0          1.0           0.0  
2       1.0          0.0           1.0  
3       1.0          0.0           1.0  
4       1.0          1.0           0.0  

[5 rows x 57 columns]

**Чтобы создать наши наборы данных для обучения и тестирования, нам нужно разделить два значения:**
- input values (which are our independent variables commonly referred to as model features or "X" in Документация TensorFlowСсылки на внешний сайт.)
- целевой вывод (наша зависимая переменная, обычно называемая целью или "y" в документации TensorFlow)

**Для наших целей мы хотим построить модель, которая предскажет, подвержен ли человек риску истощения; поэтому мы должны отделить столбцы "Выбытие" от остальных входных данных. На самом деле, поскольку данные истощения дихотомичны (одно из двух значений), нам нужно только сохранить столбец "Attrition_Yes" - мы можем игнорировать столбец "Attrition_No", потому что он избыточный. Чтобы разделить наши функции и цель, а также выполнить разделение обучения/тестирования, добавьте и запустите следующий код в своем блокноте:**

In [7]:
# Split our preprocessed data into our features and target arrays
y = attrition_df["Attrition_Yes"].values
X = attrition_df.drop(["Attrition_Yes","Attrition_No"],1).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

/Users/dang/opt/anaconda3/envs/mlenv/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  This is separate from the ipykernel package so we can avoid doing imports until


**Tеперь, когда наши данные об обучении и тестировании выделены, мы готовы создать наш StandardScaler и стандартизировать числовые функции. Добавьте и запустите следующий код в блокнот:**

In [8]:
#Create a StandardScaler instance
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


**Теперь, когда наши данные предварительно обрабатываются с помощью одноразового кодирования и стандартизации, нам, вероятно, следует выполнить проверку кишечника, чтобы убедиться, что никакие данные не были потеряны из нашего оригинального DataFrame.**

**Наконец, наши данные предварительно обработаны, разделены и готовы к моделированию. Для наших целей мы будем использовать тот же фреймворк, который мы использовали для нашей базовой нейронной сети:**
- Для нашего входного слоя мы должны добавить количество входных элементов, равное количеству переменных в нашем элементе DataFrame.
- In our hidden layers, our deep learning model structure will be slightly different—we'll add two hidden layers with only a few neurons in each layer. To create the second hidden layer, we'll add another Keras Dense class while defining our model. All of our hidden layers will use the relu activation function to identify nonlinear characteristics from the input values.
- In the output layer, we'll use the same parameters from our basic neural network including the sigmoid activation function. The sigmoid activation function will help us predict the probability that an employee is at risk for attrition.
  
**Чтобы создать нашу модель глубокого обучения, мы должны добавить и запустить следующий код:**

In [9]:
# Define the model - deep neural net
number_input_features = len(X_train[0])
hidden_nodes_layer1 = 8
hidden_nodes_layer2 = 5

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

2022-11-02 18:05:53.618664: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 8)                 448       
                                                                 
 dense_1 (Dense)             (None, 5)                 45        
                                                                 
 dense_2 (Dense)             (None, 1)                 6         
                                                                 
Total params: 499
Trainable params: 499
Non-trainable params: 0
_________________________________________________________________


**Глядя на наше резюме модели, мы видим, что количество параметров веса (весовых коэффициентов) для каждого слоя равно количеству входных значений, умноженных на количество нейронов плюс член смещения для каждого нейрона. Наш первый слой имеет 55 входных значений и умноженный на восемь нейронов (плюс восемь членов смещения для каждого нейрона), дает нам в общей сложности 448 параметров веса - множество возможностей для нашей модели найти тенденции в наборе данных.**

**Now it is time to compile our model and define the loss and accuracy metrics. Since we want to use our model as a binary classifier, we'll use the binary_crossentropy loss function, adam optimizer, and accuracy metrics, which are the same parameters we used for our basic neural network. To compile the model, add and run the following code:**

In [10]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

**В зависимости от сложности набора данных мы можем увеличить количество эпох, чтобы дать модели глубокого обучения больше возможностей для оптимизации весовых коэффициентов. Чтобы обучить нашу модель, мы должны добавить и запустить следующий код:**

In [11]:
# Train the model
fit_model = nn.fit(X_train,y_train,epochs=100)

Epoch 1/100
35/35 [==============================] - 1s 1ms/step - loss: 1578.5209 - accuracy: 0.1724
Epoch 2/100
35/35 [==============================] - 0s 1ms/step - loss: 802.3047 - accuracy: 0.1815
Epoch 3/100
35/35 [==============================] - 0s 1ms/step - loss: 121.9504 - accuracy: 0.5917
Epoch 4/100
35/35 [==============================] - 0s 1ms/step - loss: 25.1017 - accuracy: 0.7967
Epoch 5/100
35/35 [==============================] - 0s 1ms/step - loss: 11.9264 - accuracy: 0.7423
Epoch 6/100
35/35 [==============================] - 0s 1ms/step - loss: 8.5325 - accuracy: 0.7577
Epoch 7/100
35/35 [==============================] - 0s 1ms/step - loss: 7.0726 - accuracy: 0.7523
Epoch 8/100
35/35 [==============================] - 0s 1ms/step - loss: 6.4516 - accuracy: 0.7468
Epoch 9/100
35/35 [==============================] - 0s 1ms/step - loss: 6.3090 - accuracy: 0.7704
Epoch 10/100
35/35 [==============================] - 0s 1ms/step - loss: 5.8540 - accuracy: 0.7414


**Наконец, теперь, когда наша модель глубокого обучения правильно обучена, мы можем оценить производительность модели, протестировав ее прогнозные возможности на нашем наборе тестовых данных.**

In [12]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

12/12 - 0s - loss: 0.4283 - accuracy: 0.8723 - 184ms/epoch - 15ms/step
Loss: 0.42833009362220764, Accuracy: 0.8722826242446899


**Глядя на показатели эффективности нашей модели глубокого обучения, модель смогла правильно идентифицировать сотрудников, которые подвержены риску истощения примерно в 87% случаев. Учитывая, что наши входные данные включали более 30 различных переменных с более чем 1400 точками данных, модель глубокого обучения смогла создать довольно надежный классификатор.**

## Контрольно-пропускные пункты 

**Теперь, когда у нас есть данные об обучении и определена модель, мы готовы собрать и обучить нашу модель с использованием контрольно-пропускных пунктов. Чтобы использовать контрольные точки, нам нужно определить имя файла контрольной точки и путь к каталогу. Для наших целей мы помечаем наши контрольные точки по номеру эпохи и будем содержать их в их собственной папке. Это гарантирует, что наши файлы контрольных точек являются аккуратными, организованными и легко идентифицируемыми. Добавьте и запустите следующий код в наш блокнот:**

In [13]:
# Import checkpoint dependencies
import os
from tensorflow.keras.callbacks import ModelCheckpoint

# Define the checkpoint path and filenames
os.makedirs("checkpoints/",exist_ok=True)
checkpoint_path = "checkpoints/weights.{epoch:02d}.hdf5"

**После того, как мы определили структуру файла и путь к файлу, нам нужно создать объект обратного вызова для нашей модели глубокого обучения. Объект обратного вызова используется в модуле Keras для определения набора функций, которые будут применяться на определенных этапах учебного процесса. Существует ряд различных функций обратного вызова, которые могут создавать файлы журналов, принудительно останавливать обучение, отправлять сообщения о состоянии обучения или, в нашем случае, сохранять контрольные точки модели. Чтобы создать эффективный обратный вызов контрольной точки с помощью метода ModelCheckpoint, нам необходимо указать следующие параметры:**
- filepath=checkpoint_path— каталог контрольных точек и файловая структура, которые мы определили ранее
- verbose=1— мы будем уведомлены, когда контрольная точка будет сохранена в каталоге
- save_weights_only=True— сохранение полной модели каждый раз может очень быстро заполнять жесткий диск; это гарантирует, что файлы контрольных точек занимают минимальное пространство
- save_freq='epoch'— контрольные точки будут сохраняться каждую эпоху
  
**Объединив все это вместе, мы можем скомпилировать, обучить и оценить нашу модель глубокого обучения, добавив и запустив следующий код:**

In [14]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

# Create a callback that saves the model's weights every epoch
cp_callback = ModelCheckpoint(
    filepath=checkpoint_path,
    verbose=1,
    save_weights_only=True,
    save_freq='epoch')

# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=100,callbacks=[cp_callback])

# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Epoch 1/100
23/35 [==================>...........] - ETA: 0s - loss: 0.4456 - accuracy: 0.8125 
Epoch 1: saving model to checkpoints/weights.01.hdf5
35/35 [==============================] - 1s 5ms/step - loss: 0.4260 - accuracy: 0.8203
Epoch 2/100
16/35 [============>.................] - ETA: 0s - loss: 0.4075 - accuracy: 0.8262
Epoch 2: saving model to checkpoints/weights.02.hdf5
35/35 [==============================] - 0s 4ms/step - loss: 0.4134 - accuracy: 0.8240
Epoch 3/100
35/35 [==============================] - ETA: 0s - loss: 0.4036 - accuracy: 0.8267
Epoch 3: saving model to checkpoints/weights.03.hdf5
35/35 [==============================] - 0s 4ms/step - loss: 0.4036 - accuracy: 0.8267
Epoch 4/100
17/35 [=============>................] - ETA: 0s - loss: 0.4029 - accuracy: 0.8162
Epoch 4: saving model to checkpoints/weights.04.hdf5
35/35 [==============================] - 0s 3ms/step - loss: 0.3941 - accuracy: 0.8303
Epoch 5/100
30/35 [========================>.....] - ETA: 0

**After running the previous code, we have created our trained model within the Python session, as well as a folder of checkpoints we can use to restore previous model weights. Now if we ever need to restore weights, we can use the Keras Sequential model's load_weights method to restore the model weights. To test this functionality, let's define another deep learning model, but restore the weights using the checkpoints rather than training the model. Once again we must add and run the following to our notebooks:**

In [15]:
# Define the model - deep neural net
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 8
hidden_nodes_layer2 = 5

nn_new = tf.keras.models.Sequential()

# First hidden layer
nn_new.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn_new.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn_new.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Compile the model
nn_new.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

# Restore the model weights
nn_new.load_weights("checkpoints/weights.100.hdf5")

# Evaluate the model using the test data
model_loss, model_accuracy = nn_new.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

12/12 - 0s - loss: 0.3871 - accuracy: 0.8723 - 201ms/epoch - 17ms/step
Loss: 0.387145459651947, Accuracy: 0.8722826242446899


**Используя контрольные точки, мы смогли мгновенно регенерировать модель и подтвердить, что модель смогла получить точно такие же результаты.**

**В настоящее время в нашей среде ноутбуков у нас должна быть полностью обученная модель классификации, которая может предсказывать выбытие сотрудников на основе функций в наборе данных. Чтобы экспортировать обученную модель, нам нужно добавить и запустить следующий код:**

In [16]:
# Export our model to HDF5 file
nn_new.save("trained_attrition.h5")

**После запуска кода мы должны увидеть файл с именем "trained_attrition.h5", который содержит полную модель и конфигурацию. Теперь, когда мы сохранили модель, мы можем создать модель в любой момент. Давайте попробуем импортировать модель в блокнот без указания какой-либо структуры или контекста. Чтобы импортировать модель, добавьте и запустите следующий код:**

In [17]:
# Import the model to a new object
nn_imported = tf.keras.models.load_model('trained_attrition.h5')

**Наконец, мы можем проверить производительность завершенной модели на нашем тестовом наборе данных, добавив и запустив следующий код:**

In [18]:
# Evaluate the completed model using the test data
model_loss, model_accuracy = nn_imported.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

12/12 - 0s - loss: 0.3871 - accuracy: 0.8723 - 167ms/epoch - 14ms/step
Loss: 0.387145459651947, Accuracy: 0.8722826242446899


**Глядя на окончательные результаты, наша импортированная модель смогла воспроизвести те же показатели производительности, что и оригинальная модель. Используя эту же процедуру, мы можем импортировать любой тип модели Keras для оценки на наборе данных с теми же функциями.**